compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -602,102 +608,116 @@ impl UnsafeOpKind {
-                    },
-                );
+                (MutationOfLayoutConstrainedFieldRequiresUnsafeUnsafeOpInUnsafeFnAllowed { span })
+                    .into_diagnostic(&tcx.sess.parse_sess.span_diagnostic)
             }
             MutationOfLayoutConstrainedField => {
-                tcx.sess.emit_err(MutationOfLayoutConstrainedFieldRequiresUnsafe { span });
+                (MutationOfLayoutConstrainedFieldRequiresUnsafe { span })
+                    .into_diagnostic(&tcx.sess.parse_sess.span_diagnostic)
             }
             BorrowOfLayoutConstrainedField if unsafe_op_in_unsafe_fn_allowed => {
-                tcx.sess.emit_err(
-                    BorrowOfLayoutConstrainedFieldRequiresUnsafeUnsafeOpInUnsafeFnAllowed { span },
-                );
+                (BorrowOfLayoutConstrainedFieldRequiresUnsafeUnsafeOpInUnsafeFnAllowed { span })
+                    .into_diagnostic(&tcx.sess.parse_sess.span_diagnostic)
             }
             BorrowOfLayoutConstrainedField => {
-                tcx.sess.emit_err(BorrowOfLayoutConstrainedFieldRequiresUnsafe { span });
+                (BorrowOfLayoutConstrainedFieldRequiresUnsafe { span })
+                    .into_diagnostic(&tcx.sess.parse_sess.span_diagnostic)
             }
             CallToFunctionWith(did) if unsafe_op_in_unsafe_fn_allowed => {
-                tcx.sess.emit_err(CallToFunctionWithRequiresUnsafeUnsafeOpInUnsafeFnAllowed {
+                (CallToFunctionWithRequiresUnsafeUnsafeOpInUnsafeFnAllowed {
                     span,
                     function: &tcx.def_path_str(*did),
-                });
+                })
+                .into_diagnostic(&tcx.sess.parse_sess.span_diagnostic)
             }
             CallToFunctionWith(did) => {
-                tcx.sess.emit_err(CallToFunctionWithRequiresUnsafe {
-                    span,
-                    function: &tcx.def_path_str(*did),
-                });
+                (CallToFunctionWithRequiresUnsafe { span, function: &tcx.def_path_str(*did) })
+                    .into_diagnostic(&tcx.sess.parse_sess.span_diagnostic)
             }
+        };
+
+        let note_non_inherited = tcx.hir().parent_iter(hir_context).find(|(id, node)| {
+            if let hir::Node::Expr(block) = node
+                && let hir::ExprKind::Block(block, _) = block.kind
+                && let hir::BlockCheckMode::UnsafeBlock(_) = block.rules
+            {
+                true
+            }
+            else if let Some(sig) = tcx.hir().fn_sig_by_hir_id(*id)
+                && sig.header.is_unsafe()
+            {
+                true
+            } else {
+                false
+            }
+        });
+        if let Some((id, _)) = note_non_inherited {
+            let span = tcx.hir().span(id);
+            err.span_label(
+                tcx.sess.source_map().guess_head_span(span),
+                crate::fluent_generated::mir_build_unsafe_not_inherited,
+            );
         }
```

Comment:
Can this be done by adding a `UnsafeNotInheritedNote` in errors structs, and add an `Option<UnsafeNotInheritedNote>` in each of the emitted error types?

Reply:
Yep - updated the PR.